packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -1335,7 +1335,8 @@ function applyAttributes(ctx: ParserCtx, parsedAttr: ParsedAttribute, element: B
 
         // disallow attr name which combines underscore character with special character.
         // We normalize camel-cased names with underscores caMel -> ca-mel; thus sanitization.
-        if (name.match(/_[^a-z0-9]|[^a-z0-9]_/)) {
+        // Note 1: underscore followed by a hyphen is valid to accomodate property names like thirsty_Camel -> thirsty_-camel
+        if (name.match(/_[^a-z0-9-]|[^a-z0-9]_/)) {
```

Comment:
This is the only code change in this PR, the rest are all tests.

Reply:
For a follow-up PR: What is the value of restricting the following pattern `/_[^a-z0-9-]|[^a-z0-9]_/`? The only remaining test case is `under-_hyphen`. This name is indeed ambiguous, that said it also blocks other names like `foo__bar` which is a valid JavaScript identifier.

Reply:
>What is the value of restricting the following pattern /_[^a-z0-9-]|[^a-z0-9]_/?

Yeah, that is a great question. I do not see much value in having this restriction IMO. I added the other test cases(valid-non-alphanumeric/actual.html, valid-hyphen-attribute/actual.html) to establish the current treatment of non alpha numeric characters. 

If non-alpha numeric characters are allowed in an attribute name in general, why only restrict is in combination with an underscore(`_`), at the start and end? The follow up work is to explore more about these restrictions and take the next steps as part of W-12021709.

Reply:
Followup PR https://github.com/salesforce/lwc/pull/3152